### ISSUE TITLE
`mypy` doesn't understand that `*args: P.args` implies that `args` is a `tuple`

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#19663]
```python
from typing import Callable

def as_tuple[*Ts](*args: *Ts) -> tuple[*Ts]: return args
def tuple_identity[*Ts](t: tuple[*Ts]) -> tuple[*Ts]: return t
def tuple_identity2[T: tuple](t: T) -> T: return t


def test_paramspec[**P](
    dummy: Callable[P, None],  # ensure P is bound
    /,
    *args: P.args,
    **kwargs: P.kwargs,
) -> None:
    reveal_type(args)                   # N: "P.args`-1" 
    reveal_type( (*args,) )             # N: "builtins.tuple[P.args`-1, ...]"
    reveal_type( tuple(args) )          # N: "builtins.tuple[builtins.object, ...]"
    reveal_type(as_tuple(*args))        # N: "builtins.tuple[P.args`-1, ...]"
    reveal_type(tuple_identity(args))   # N: "builtins.tuple[Never, ...]"
                                        # E: [arg-type]
    reveal_type(tuple_identity2(args))  # N: "P.args`-1" ✅

    if isinstance(args, tuple):
        pass
    else:
        reveal_type(args)  # false negative [warn-unreachable]
```

https://mypy-play.net/?mypy=latest&python=3.12&flags=warn-unreachable&gist=3eabd8f4ee599e272934c94080f02bd8

Ideally, all the `reveal_types` should show the same result (or raise errors if considered misuse of `ParamSpec`[^1]), and the `else`-branch should trigger an `unreachable` warning.

[^1]: For instance, `pyright` says `as_tuple(*args)` is illegal. Code sample in [pyright playground](https://pyright-play.net/?reportUnreachable=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgFC0Am1wURAzgPowCuCNA2gCoAKmwC6ACkFEQaNgC4oItgEooAWgB8sXgOVjFIajxApWsto2Y6%2B1DkiYoYSeENGSYinrbfi1Wmz13Q2NuU1grFm8ae0dneAAmfmEvXWoPRWF-bRSoIxMzGHpI2Go2GA4EGSIINgRqAGMhQQAFSVooTqgGbggIOEViMkoBFoAaKAA5MBR0iagAYihqFDYw6igWzDYoCjBuFAYOroB6MePO6QtFFoA6GTlzrqVBAGsAdweFTduPr-PslMZtR5Bc8tQAG7UUhcRDUCRfNTPZEorpLSaKABEdy%2BAAN1ABGTFQMFGKEw%2BD1CRQKT-NRI1HPdFYijcJAkZyrW7Raj8HEWfEEia3EViTGkyHQkiwqmBeGIqAM1HMqCY1nszlsblpfjqjmoLVgCgAK0aMGFovFzzJUpl8s4PNpFhUSqZUxZbP1XJ5fPuAsJFtuYol5OllPhPNiK3icARzoZKrVns12p8k0lIEDwcZOdzec6SwAoop%2BA91OGxCHbeGJJGHNGXHAEnG5C6C%2B7Vfy5ILiYBQcnozyQLCQbANMCIKAa8osEx5KlBjKqbEsz2oJDYILB1slFLhLdU7eApA3UFmaCIzihUH4n1M6gORiIDQAFiN0kA)

**Comment by sterliakov** [comment]
We shouldn't replicate pyright behavior here, it is not a type error but a weirdly phrased "not supported" message ([discussion](https://github.com/microsoft/pyright/discussions/5049)). That usage is perfectly fine and (IMO) should produce a `tuple[object, ...]`.
